### PR TITLE
renderer_vulkan: Spirv support glsl  legacy attribute

### PR DESCRIPTION
**Pull request by vonchenplus**
This PR aims to support the gl_Color and gl_TexCoord attributes of glsl in spirv.

Solved similar DQ3 black screen issue in vulkan mode.

### CODE REVIEW

File: src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp [hunk 1]
```diff
@@ -85,6 +91,14 @@ std::optional<OutAttr> OutputAttrPointer(EmitContext& ctx, IR::Attribute attr) {
         const Id element_id{ctx.Const(element)};
         return OutputAccessChain(ctx, ctx.output_f32, ctx.output_position, element_id);
     }
+    case IR::Attribute::ColorFrontDiffuseR:
+    case IR::Attribute::ColorFrontDiffuseG:
+    case IR::Attribute::ColorFrontDiffuseB:
+    case IR::Attribute::ColorFrontDiffuseA: {
+        const u32 element{static_cast<u32>(attr) % 4};
```
Comment (v1993) [inline src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp:98]: How about using `TxtCoordAttributeElement` here?

Comment (vonchenplus) [reply src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp:99]: My naming scheme here is the same as IR::Attribute::PositionX

Comment (v1993) [reply src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp:99]: I'm not talking about naming here. It just looks like `TxtCoordAttributeElement` exists specifically to do this conversion (it's the same code for sure).

Comment (vonchenplus) [reply src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp:99]: Oh, yes, I misunderstood. Actually calculating AttributeElement is the same in several places, and I can't quite pinpoint this thing, so for now I'll probably leave this piece alone for now.

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 2]
```diff
@@ -1227,17 +1228,15 @@ void EmitContext::DefineInputs(const IR::Program& program) {
         loads[IR::Attribute::TessellationEvaluationPointV]) {
         tess_coord = DefineInput(*this, F32[3], false, spv::BuiltIn::TessCoord);
     }
+    std::queue<u32> ununsed_location;
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:1231]: Rather than use a std::queue, I'd prefer the use of a std::bitset to represent the availability of all generics location.

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 3]
```diff
@@ -1227,17 +1228,15 @@ void EmitContext::DefineInputs(const IR::Program& program) {
...
-        if (!loads.Generic(index)) {
-            continue;
-        }
-        if (input_type == AttributeType::Disabled) {
+        if (!runtime_info.previous_stage_stores.Generic(index) || !loads.Generic(index) ||
+            input_type == AttributeType::Disabled) {
+            ununsed_location.push(static_cast<u32>(index));
             continue;
         }
+
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:1239]: Unneeded newline

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 4]
```diff
@@ -1227,17 +1228,15 @@ void EmitContext::DefineInputs(const IR::Program& program) {
...
         const AttributeType input_type{runtime_info.generic_input_types[index]};
-        if (!runtime_info.previous_stage_stores.Generic(index)) {
-            continue;
-        }
-        if (!loads.Generic(index)) {
-            continue;
-        }
-        if (input_type == AttributeType::Disabled) {
+        if (!runtime_info.previous_stage_stores.Generic(index) || !loads.Generic(index) ||
+            input_type == AttributeType::Disabled) {
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:1235]: I'd revert this change and keep the conditions separated, seeing as the conditions are not entirely related.

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 5]
```diff
@@ -1313,9 +1334,34 @@ void EmitContext::DefineOutputs(const IR::Program& program) {
         viewport_mask = DefineOutput(*this, TypeArray(U32[1], Const(1u)), std::nullopt,
                                      spv::BuiltIn::ViewportMaskNV);
     }
+    std::queue<u32> ununsed_location;
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:1337]: ditto about the bitset

File: src/shader_recompiler/frontend/ir/attribute.h [hunk 6]
```diff
@@ -222,8 +222,15 @@ enum class Attribute : u64 {
     FrontFace = 255,
 };
 
+constexpr size_t NUM_FIXEDFNCTEXTURE = 10;
```
Comment (ameerj) [inline src/shader_recompiler/frontend/ir/attribute.h:225]: Since this is currently only used in the `spirv/emit_context.cpp`, it would be more appropriate to declare it in the anonymous namespace of that file

File: src/shader_recompiler/frontend/ir/attribute.h [hunk 7]
```diff
@@ -222,8 +222,15 @@ enum class Attribute : u64 {
...
 constexpr size_t NUM_GENERICS = 32;
 
+[[nodiscard]] bool IsFixedFncTexture(Attribute attribute);
+
+[[nodiscard]] u32 FixedFncTextureAttributeIndex(Attribute attribute);
+
+[[nodiscard]] u32 FixedFncTextureAttributeElement(Attribute attribute);
+
```
Comment (ameerj) [inline src/shader_recompiler/frontend/ir/attribute.h:233]: These should be moved to the anonymous namespace of `spirv/emit_spirv_context_get_set.cpp`

**Review by ameerj — COMMENTED** [review]
Few nits

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 8]
```diff
@@ -427,6 +427,18 @@ Id DescType(EmitContext& ctx, Id sampled_type, Id pointer_type, u32 count) {
         return pointer_type;
     }
 }
+
+constexpr size_t NUM_FIXEDFNCTEXTURE = 10;
+
+size_t FindFistUnUsedLocation(const auto& used_location) {
+    if (used_location.all()) {
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:434]: Since this case is unlikely, and may need to iterate the entire bitset to evaluate, we can save ourselves from looping the bitset twice by refactoring this function to be a for loop through all indices, returning early if a free index is found, otherwise throwing at the end.

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 9]
```diff
@@ -427,6 +427,18 @@ Id DescType(EmitContext& ctx, Id sampled_type, Id pointer_type, u32 count) {
...
+        throw RuntimeError("Unable to get an unused location");
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:435]:
```suggestion
        throw RuntimeError("Unable to get an unused location for legacy attribute");
```

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 10]
```diff
@@ -1227,6 +1239,7 @@ void EmitContext::DefineInputs(const IR::Program& program) {
         loads[IR::Attribute::TessellationEvaluationPointV]) {
         tess_coord = DefineInput(*this, F32[3], false, spv::BuiltIn::TessCoord);
     }
+    std::bitset<IR::NUM_GENERICS> used_location;
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:1242]:
```suggestion
    std::bitset<IR::NUM_GENERICS> used_locations{};
```

File: src/shader_recompiler/backend/spirv/emit_context.cpp [hunk 11]
```diff
@@ -1263,6 +1277,22 @@ void EmitContext::DefineInputs(const IR::Program& program) {
...
+    if (loads.AnyComponent(IR::Attribute::ColorFrontDiffuseR)) {
+        size_t location = FindFistUnUsedLocation(used_location);
+        used_location.set(location);
+        const Id id{DefineInput(*this, F32[4], true)};
+        Decorate(id, spv::Decoration::Location, location);
+        input_front_color = id;
+    }
+    for (size_t index = 0; index < NUM_FIXEDFNCTEXTURE; ++index) {
+        if (loads.AnyComponent(IR::Attribute::FixedFncTexture0S + index * 4)) {
+            size_t location = FindFistUnUsedLocation(used_location);
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_context.cpp:1289]: I'd suggest keeping a running index of the available position in the bitset, and pass that in as an argument to this function to find the next available position, rather than repeating the work of looking for the position from index 0 everytime.

File: src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp [hunk 12]
```diff
@@ -43,6 +43,25 @@ Id AttrPointer(EmitContext& ctx, Id pointer_type, Id vertex, Id base, Args&&...
...
 }
 
+bool IsFixedFncTexture(IR::Attribute attribute) {
+    return attribute >= IR::Attribute::FixedFncTexture0S &&
+           attribute <= IR::Attribute::FixedFncTexture9Q;
+}
+
+u32 FixedFncTextureAttributeIndex(IR::Attribute attribute) {
+    if (!IsFixedFncTexture(attribute)) {
+        throw InvalidArgument("Attribute is not fixedfnctexture {}", attribute);
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp:53]:
```suggestion
        throw InvalidArgument("Attribute {} is not a FixedFncTexture", attribute);
```

File: src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp [hunk 13]
```diff
@@ -43,6 +43,25 @@ Id AttrPointer(EmitContext& ctx, Id pointer_type, Id vertex, Id base, Args&&...
...
+    }
+    return (static_cast<u32>(attribute) - static_cast<u32>(IR::Attribute::FixedFncTexture0S)) / 4u;
+}
+
+u32 FixedFncTextureAttributeElement(IR::Attribute attribute) {
+    if (!IsFixedFncTexture(attribute)) {
+        throw InvalidArgument("Attribute is not fixedfnctexture {}", attribute);
+    }
+    return static_cast<u32>(attribute) % 4;
```
Comment (ameerj) [inline src/shader_recompiler/backend/spirv/emit_spirv_context_get_set.cpp:62]:
```suggestion
    return static_cast<u32>(attribute) % 4u;
```

File: src/shader_recompiler/backend/spirv/emit_context.cpp (hunk 11 above)
Comment (vonchenplus) [reply src/shader_recompiler/backend/spirv/emit_context.cpp:1289]: The reason for this is that the use of broken locations is not continuous, for example, only 0 and 2 were used before, and the location 1 was left vacant

Comment (ameerj) [reply src/shader_recompiler/backend/spirv/emit_context.cpp:1289]: This won't be an issue. In your example, the first call to the function will check location 0, find it is not available, check 1, return that as the index of an available location. The next time the function is called, there is no reason to check indices 0 and 1 again as they have already been checked and determined to be used, so you begin searching from index 2.

Comment (vonchenplus) [reply src/shader_recompiler/backend/spirv/emit_context.cpp:1289]: Well, I guess it's not a performance bottleneck here, doing so leads to an increase in code complexity?

**Review by ameerj — COMMENTED** [review]
Last few nits

**Review by bunnei — APPROVED** [review]

**Comment by bunnei** [comment]
LGTM, thank you!
